### ISSUE TITLE
Questions about token shifting in the original data

### ISSUE THREAD

**Issue by HelloWorldLTY** [gerstung-lab/Delphi#27]
Hi, I successfully trained my own data based on delphi framework, but one thing makes me feel confused is the usage of death token and related token shift. In the simulation data , the max id number is 1268, but in the labels.txt file, the max token is 1270. I see that the model also has vocab size as 1270. Does this mean simulation data do not contain death tokens?

I test it again with my own data, but I meet error caused by the mismatch between log prob head output and observed label. The fact is I must increase the vocab_size by 2, for example, 6740->6742, thought my max token is still 6740, which is death.

Therefore, I want to figure out what is the reason of this setting. Thanks a lot.

Here is my labels.txt file's top content:

<img width="936" height="501" alt="Image" src="https://github.com/user-attachments/assets/f48c0a8d-17bc-43a4-abc2-faae55aa3707" />

Thanks a lot.

**Comment by yozhikoff** [comment]
Hi @HelloWorldLTY !

We have a description of how to prepare a new dataset here:
https://github.com/gerstung-lab/Delphi/blob/main/data/README.md

Let me give a concrete example, because I understand this is all a bit confusing.
Let's have diseases A, B, death D and padding token P

In the binary files like `train.bin`, the mapping to the token index would be 
`A: 1, B: 2, D:3`
The padding and no event tokens are not recorded there
Basically, you only need to ensure that the minimal index there is 1
This enumeration is only used in the bin files and nowhere else

When the `get_batch()` function prepares the data, it shifts the tokens by one 
`token_id = token_id + 1` and how the mapping is
`A: 2, B: 3, D:4`
Now, we use the newly created room to add padding tokens P and no event tokens N
`P:0, N:1, A: 2, B: 3, D:4`
This mapping is used everywhere - labels.txt, input and outputs of the transformer, etc
In our example, the death logit can be accessed as `logits[id_in_batch, id_within_traj, 4]`

The best way to look for the mapping used in our model is here https://github.com/gerstung-lab/Delphi/blob/main/delphi_labels_chapters_colours_icd.csv, as it has a dedicated index column. Death there is 1269, which is 1268 (max token in data) + 1.

The labels.txt file makes it look as if death is 1270, but this is an illusion created by the fact that GitHub enumerates lines from 1, while indices of tokens are enumerated from 0. These line numbers aren't even in the file itself. 

So, to wrap up:
- Start your data in .bin files with 1
- There will be +1 shift in the dataloader, but it happens automatically - just don't forget about it when creating labels for your data
- Model vocabulary is indeed max_token_id + 2

I hope it helps!

**Comment by HelloWorldLTY** [comment]
Makes sense. Now I totally understand and thank you for your explaniations!

**Comment by HelloWorldLTY** [comment]
Sorry to distrub, after reading the document, I have some new questions:

```
patient_id should be a unique identifier for each patient. It is used exclusively for identifying which records belong to the same patient and it is not used in the training. The entries should be consecutive entries, meaning that first in the array come all the rows related to the first patient, then all the rows related to the second patient and so on. patient_age is the age of the patient in days at the time of recording the disease. token_id is some unique identifier for the data record, a uint32 number, starting from 1 (0 is reserved for no event tokens). In Delphi-2M, it could be a disease event, a lifestyle token or sex token. All the tokens are processed by the model in a unified way (however, there are some tokens that are not used for calculating the gradients - see below for more details).
```

Here you mentioned: "starting from 1 (0 is reserved for no event tokens)", but in the simulation data, 0 is actually for padding token. Did I miss anything? I think we should minus 1 for all mapped token from labels.csv, to ensure the token is consistent (I refer the simulation data example, here it seems every event maps to the place where index-1 from labels.csv)? 

This is what my labels.csv looks like:

<img width="936" height="501" alt="Image" src="https://github.com/user-attachments/assets/fbb5e2ea-9d73-4f19-8540-132490615811" />

Also, in the simulation example, you have provided example trajectory. Do we also need to adjust that event token? e.g., +1 for mapping?

My last question is for death token. Do we need to consider special processing approach for the dealth token?

I am sorry for multiple questions... As my trajectory always does not produce any no event token, I beat there are some issues I misunderstand.

**Comment by yozhikoff** [comment]
> Here you mentioned: "starting from 1 (0 is reserved for no event tokens)", but in the simulation data, 0 is actually for padding token. Did I miss anything?

No, it's my bad! It's a typo in the document. 0 is indeed padding, 1 is no event. I will fix it, thanks for spotting!

> I refer to the simulation data example, here it seems every event maps to the place where index-1 from labels.csv

If you mean the simulation data in the .bin files - yes, it's minus 1 wrt the label files. 
In general, I suggest looking at the get_batch() function here:
https://github.com/gerstung-lab/Delphi/blob/231a88bb0fc2f63e2296d70def74feccb01ff6b7/utils.py#L26
All the conversions are happening there, so it may be helpful to understand the underlying logic

> Also, in the simulation example, you have provided example trajectory. Do we also need to adjust that event token? e.g., +1 for mapping?

Could you clarify which example you mean?

> My last question is for death token. Do we need to consider special processing approach for the dealth token?

For sampling novel trajectories, one generally wants to stop after encountering a death token. Aside from that, there is no special treatment in preprocessing.

  > I am sorry for multiple questions... As my trajectory always does not produce any no event token, I beat there are some issues I misunderstand.

Please feel free to ask more questions if needed!
We can also have a short call to clarify the caveats of data processing

**Comment by HelloWorldLTY** [comment]
Thank you, that is really helpful. I actually successfully reproduce most of the content in the demo, thank you for the great work. 

```
Could you clarify which example you mean?
```

I have found a solution is to minus 1 when constructing the training and validation data for the event label, and after the +1 operation in the training data, the order will not be affected.

I will investigate the codes for embedding generation for more details. Thanks a a lot again!
